pkg/yttlibrary/regexp.go

```diff
@@ -47,3 +48,67 @@ func (b regexpModule) Match(thread *starlark.Thread, f *starlark.Builtin, args s
 
 	return starlark.Bool(matched), nil
 }
+
+func (b regexpModule) Replace(thread *starlark.Thread, f *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	if args.Len() != 3 {
+		return starlark.None, fmt.Errorf("expected exactly 3 arguments")
+	}
+
+	pattern, err := core.NewStarlarkValue(args.Index(0)).AsString()
+	if err != nil {
+		return starlark.None, err
+	}
+
+	re, err := regexp.Compile(pattern)
+	if err != nil {
+		return starlark.None, err
+	}
+
+	source, err := core.NewStarlarkValue(args.Index(1)).AsString()
+	if err != nil {
+		return starlark.None, err
+	}
+
+	repl := args.Index(2)
+	switch repl.(type) {
```

Comment:
This implementation keeps a single `regexp.replace` method which can take either a string or lambda as the third param. To client users, I think this makes sense:

1. In both cases the third param is the "repl" argument, i.e. the thing to replace matches with, just expressed differently.
2. Starlark, being based on Python, is a dynamic language, so no need to provide methods with different names when they accept similar arguments with just slightly different signatures.

However:

3. From skimming through the API, I think this may be the first time that the ytt library adds a method which requires a dynamic type check like this? If so: are we ok with that?
4. This check works but I'm not 100% certain there isn't a more canonical way to use the starlark-go API. I'll do some further digging when I have time, but figured I'd push what I have so far to get feedback.

Reply:
> From skimming through the API, I think this may be the first time that the ytt library adds a method which requires a dynamic type check like this? If so: are we ok with that?

we do something similar for some overlay keyword args e.g. https://github.com/k14s/ytt/blob/develop/pkg/yttlibrary/overlay/match_annotation_expects_kwarg.go#L91. using Callable interface is better for type check.